src/test/java/net/sf/jabref/importer/fileformat/BibtexParserTest.java

```diff
@@ -32,219 +104,682 @@ public void testParseReader() throws IOException {
+
+        Collection<BibtexEntry> c = result.getDatabase().getEntries();
+        Assert.assertEquals(1, c.size());
+
+        BibtexEntry e = c.iterator().next();
+        Assert.assertEquals(BibtexEntryTypes.ARTICLE, e.getType());
+        Assert.assertEquals("test", e.getCiteKey());
+        Assert.assertEquals(2, e.getFieldNames().size());
+        Assert.assertEquals("1-4~#nov#", e.getField("date"));
+    }
+
+    @Test
+    public void parseReturnsEmptyListIfNoEntryRecognized() throws IOException {
+        ParserResult result = BibtexParser.parse(new StringReader(
+                    "  author = {Crowston, K. and Annabi, H. and Howison, J. and Masango, C.},"
+                            + "\n"
+                            + "  title = {Effective work practices for floss development: A model and propositions},"
+                            + "\n"
+                            + "  booktitle = {Hawaii International Conference On System Sciences (HICSS)},"
+                            + "\n" + "  year = {2005}," + "\n" + "  owner = {oezbek}," + "\n"
+                            + "  timestamp = {2006.05.29}," + "\n"
+                            + "  url = {http://james.howison.name/publications.html}" + "\n" + "}))"
+                ));
+        Collection<BibtexEntry> c = result.getDatabase().getEntries();
+        Assert.assertEquals(0, c.size());
+    }
+
+    @Test
+    public void parseReturnsEmptyListIfNoEntryExistent() throws IOException {
+
+        ParserResult result = BibtexParser.parse(new StringReader(
+                "This was created with JabRef 2.1 beta 2." + "\n" + "Encoding: Cp1252" + "\n"));
+        Collection<BibtexEntry> c = result.getDatabase().getEntries();
+        Assert.assertEquals(0, c.size());
+    }
+
+    @Test
+    public void parseRecognizesDuplicateBibtexKeys() throws IOException {
+
+        ParserResult result = BibtexParser
+                .parse(new StringReader("@article{canh05," + "  author = {Crowston, K. and Annabi, H.},\n"
+                        + "  title = {Title A}}\n" + "@inProceedings{canh05," + "  author={Norton Bar}}"));
+
+        String[] duplicateKeys = result.getDuplicateKeys();
+        Assert.assertEquals(1, duplicateKeys.length);
+        Assert.assertEquals("canh05", duplicateKeys[0]);
+    }
+
+    @Test
+    public void parseWarnsAboutEntryWithoutBibtexKey() throws IOException {
+
+        ParserResult result = BibtexParser.parse(new StringReader("@article{,author={Ed von Test}}"));
+
+        Assert.assertTrue(result.hasWarnings());
+
+        Collection<BibtexEntry> c = result.getDatabase().getEntries();
+        Assert.assertEquals(1, c.size());
+
+        BibtexEntry e = c.iterator().next();
+        Assert.assertEquals(BibtexEntryTypes.ARTICLE, e.getType());
+        Assert.assertEquals(null, e.getCiteKey());
+        Assert.assertEquals(2, e.getFieldNames().size());
+        Assert.assertEquals("Ed von Test", e.getField("author"));
```

Comment:
What is the canonical bibtex format? Do you mean, that the `BibtexEntry` should be converted to a bibtex string? I strongly oppose this change as then the parser tests could fail for two different reasons: the parser logic is wrong or the export to bibtex has a flaw.